stor/dx.py

```diff
@@ -113,96 +165,727 @@ def temp_url(self, lifetime=300, filename=None):
+        with _propagate_dx_exceptions():
+            file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                       project=self.canonical_project)
+            return file_handler.get_download_url(
+                duration=lifetime,
+                preauthenticated=True,
+                filename=filename,
+                project=self.canonical_project
+            )[0]
 
     @property
     def resource(self):
+        """Returns the resource as a ``PosixPath`` object or None.
+
+        The returned resource is defined relative to a project.
+        For virtual paths like DXPath('dx://project:dir/file'), the resource
+        is PosixPath('/dir/file').
+        For canonical paths like
+        DXPath('dx://project-FJq288j0GPJbFJZX43fV5YP1:file-FKzjvV80jfv8pFzQ0KfbYY6B'),
+        the resource is PosixPath('/file-FKzjvV80jfv8pFzQ0KfbYY6B').
+
+        The operation parses the object string, and hence no API request is performed.
+        """
         parts = self._get_parts()
         joined_resource = '/'.join(parts[1:]) if len(parts) > 1 else None
         return self.parts_class('/'+joined_resource) if joined_resource else None
 
     def dirname(self):
+        """Returns directory name of path. Returns self if path is a project.
+
+        For canonical paths, returns the project ID, since this parses the object
+        as string without making API calls to the server.
+        """
         if not self.resource:
             return self
         else:
-            return super(DXPath, self).dirname()
+            parts = self._get_parts()
+            if len(parts) == 2:  # paths like ('dx://proj:file') need different logic
+                parts[0] += ':'
+                new_path = '/'.join(parts)
+                return self.path_class(self.drive + self.path_module.dirname(new_path))
+            else:
+                return super(DXPath, self).dirname()
+
+    @property
+    def name(self):
+        """Returns File or folder name of the path. Empty string for projects or folders
+        with trailing slash.
 
-    def download_objects(self):  # may not need it
-        raise NotImplementedError
+        Makes no API calls to server, canonical paths are treated normally, and the basename
+        of the path is returned.
+        """
+        parts = self._get_parts()
+        if len(parts) == 2 and parts[1]:  # paths like ('dx://proj:file') need different logic
+            parts[0] += ':'
+            new_path = '/'.join(parts)
+            return self.parts_class(self.path_module.basename(new_path))
+        else:
+            return super(DXPath, self).name
 
     def remove(self):
-        raise NotImplementedError
+        """Removes a single object from DX platform
 
+        Raises:
+            ValueError: The path is invalid.
+        """
+        if not self.resource:
+            raise ValueError('DXPath must point to single data object to call remove')
```

Comment:
nit: `DXPath.remove() can only be called on single objects`

Reply:
Changed.